File: JavaScript/Easy/Q1071-GreatestCommonDivisorOfStrings.js
/**
 * @param {string} str1
 * @param {string} str2
 * @return {string}
 */
var gcdOfStrings = function (str1, str2) {
    if (str1 + str2 !== str2 + str1) return "";

    let min = Math.min(str1.length, str2.length);
    for (let i = min; i > 0; i--) {
        if (str1.length % i === 0 && str2.length % i === 0) {
            return str1.substring(0, i);
        }
    }
};

const gcdOfStrings = (str1, str2) => {
    if (str1 + str2 !== str2 + str1) return "";

    function gcd(len1, len2) {
        return len2 === 0 ? len1 : gcd(len2, len1 % len2);
    }

    return str1.substring(0, gcd(str1.length, str2.length));
};

File: Python/Leetcode75/Easy/String/Q1071-GreatestCommonDivisorOfStrings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        gcds = ""
        n = min(len(str1), len(str2))

        for i in range(n):
            if str1.replace(str1[:i+1], "") == "" and str2.replace(str1[:i+1], "") == "":
                gcds = str1[:i+1]
            elif str1.replace(str2[:i+1], "") == "" and str2.replace(str2[:i+1], "") == "":
                gcds = str2[:i+1]
        return gcds
    
    def gcdOfStringsOptimized(self, str1: str, str2: str) -> str:
        if str1 + str2 != str2 + str1:
            return ""
        
        def gcd(dividend, divisor):
            if divisor == 0:
                return dividend
            return gcd(divisor, dividend % divisor)
        
        return str1[:gcd(len(str1), len(str2))]
